Automation (3/3): livepatch extraction

fgyanz opened the pull request:
This PR depend on: #154, #156 and [klp-ccp macro fix](https://github.com/fgyanz/klp-ccp/tree/headers-fix). The latter being a branch that:
- Removes the prologue of low-level macros added to every generated livepatch file. The issue was with the header inclusion heuristics, so now klp-ccp will just include all the headers instead.
- Export a `patched_funcs` file with all the livepatched functions (excluding inlined ones).

Based on the above information, this PR adds the following improvements:
- Declaring the klpp functions in the livepatch header.
- Declaring needed structs in the header.
- Removing the 'static' keyword from the klpp functions definition.
- Exporting the right symbols in "patched_funcs.csv".
- Better support for multi-file livepatches. Wire-up the inits() and cleanups() in the main livepatch.c file. (#11 )
- Adds a worktree for the kgraft-patches repository and a `kgraft` library for interacting with it.
- In the `push` command, instead of cloning the kgraft-patches repository for each
  codestream, use instead the `kgraft` library and have a branch for each codestream.
- Add a new 'commit' command for commiting the extracted livepatches to the kgraft-patches repository.

Take for instance CVE  `2024-53164`. Up until now, the livepatch would have been generated by running:
```
1)   klp-build setup --cve 2024-53164 -n bsc1246019
                   --conf CONFIG_NET_SCH_CAKE --mod-file-funcs sch_cake net/sched/sch_cake.c cake_drop
                   --conf CONFIG_NET_SCH_CHOKE --mod-file-funcs sch_choke net/sched/sch_choke.c choke_drop_by_idx

2) klp-build extract -n bsc1246019 --apply-patches
3) manually create the branches in kgraft-patches and edit the livepatches.
4) klp-build push -n bsc1246019
```

The proposed changes would just require:
```
1) klp-build setup --cve 2024-53164 -n bsc1246019
2) klp-build extract -n  bsc1246019
3) klp-build commit -n bsc1246019 (optionally)
4) klp-build push -n bsc1246019
```

fgyanz committed:
plugins/extract: Apply the patches with git

The current implementation uses quilt to apply the patches,
and although it works in most cases, it lacks the flexibility
offered by git. In the past it made sense to use quilt due to
several limitations in the tooling, but that's not the case anymore.
Keep patch(1) as a fallback in case git-am(1) fails.

From now on, during 'extract' each codestream will have its own
working branch, making it easier to track changes in case of errors.
From a high level the process will remain the same as before:
0. run extract()
1. remove patches
2. apply patches
3. run klp-ccp
4. remove patches

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
plugins/extract: Retrieve the livepatched functions

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
klplib/templ: Integrate klpp symbols into templates

The 'extract' command is finally aware of the functions klp-ccp
decided to livepatch (aka 'klpp_symbols'). As such, klp-build can
now automate all steps the users had to manually do before:

- Declaring the klpp functions in the livepatch header.
- Declaring needed structs in the header.
- Exporting the right symbols in "patched_funcs.csv".

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
plugins/extract: Cleanup generated livepatch

Now that klp-ccp does not generate the low-level macro prelude
(see https://github.com/SUSE/klp-ccp/issues/8), cleanup some
of the small klp-ccp leftovers.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
klplib/templ: Support multi-file init/cleanup

Wire-up the inits() and cleanups() in the main livepatch.c file.
This is meant only for the non-ibt codestreams. The ibt ones
don't need a livepatch.c file.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
klplib/utils: Unclassify codestream groups

Provide functions for extracting the codestreams included in
a grouped-format string e.g: "15.5u12-16, 15.6rtu8-10"

Quite useful specially for knowing which codestreams are included
in the "ccp/groups" file.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
klplib/kgraft: Add kgraft-patches library

Add a library for interacting with the kgraft-patches repository
in a transparent way. Internally, this library create a git worktree
instead of directly using the main one.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
plugins/commit: Add new 'commit' command

Provide a new command for commiting the extracted livepatches
to the kgraft-patches repository. The branches are created as
per indicated in the ccp/groups file. The default policy forbids
from modifying already existing branches, to avoid erasing any
user-changes. A "--force" option is provided to overwrite them.

A basic use-case would be as follow:
$ klp-build extract -n bsctest
$ klp-build commit -n bsctest
... # do some changes to the livepatches
$ klp-build push -n bsctest

Should the user need to extract again the livepatch, a 'force'
would be required:
$ klp-build extract -n bsctest
$ klp-build commit -n bsctest --force
...
$ klp-build push -n bsctest

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
plugins/push: Use kgraft library

Instead of cloning the kgraft-patches repository for each
codestream, use instead the kgraft library and have a branch
for each codestream.

Attempt a soft-commit before running push().

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

fgyanz committed:
plugins/extract: Apply patches by default

Reverse the default policy: Apply the patches always before extraction
unless specified by the user with the "--no-patches" option.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

vmezzela commented on klpbuild/klplib/templ.py:
```diff
@@ -482,24 +545,29 @@ def __generate_lp_file(lp_name, lp_path, cs, src_file, out_name):
         lp_inc_dir = Path("non-existent")
     else:
         fdata = cs.files[str(src_file)]
+        mod = cs.get_file_mod(src_file)
```
Same in `pathced_funcs.csv`

fgyanz replied:
yep, is that a big deal?  AFAIK it should work all the same, no?

vmezzela replied:
At the time I added  this comment I didn't know it would have worked anyway :D 
But actually you are right, so no issue.

vmezzela replied:
Hello, turns out this can create issues at runtime: it makes the module notifier return early without solving some externalizations.

Hence we should fix it and use only the name of the module rather than the path.

fgyanz commented:
Thanks @vmezzela ! you are right. I just stumbled upon this same issue in a testing environment. I'll fix it asap :)

vmezzela commented on klpbuild/plugins/extract.py:
```diff
@@ -380,32 +380,74 @@ def group_equal_files(lp_name, working_cs):
+            raise RuntimeError(f"Failed to delete branch {bname}:\n{err.stderr}\n")
 
 
+<<<<<<< Updated upstream
```
leftover?

fgyanz replied:
~~humm I think you are using and outdated version.~~

vmezzela replied:
mmh It's the last version of the commit "plugins/extract: Apply the patches with git"

fgyanz replied:
My bad, rebase chaos. Fixed

vmezzela commented on klpbuild/plugins/extract.py:
```diff
@@ -380,32 +380,74 @@ def group_equal_files(lp_name, working_cs):
 def apply_all_patches(lp_name, cs, apply_patches):
+=======
+
+def apply_patch(patch, sdir):
```
can't we just use `patch` here? why do we need to commit?

fgyanz replied:
Ah, sorry, I failed to clarify that point in the comments. There are actually several reasons:
- `git-am` and `patch` have different algorithms for resolving conflicts and applying patches. It's true that `patch` will work in most cases better that `git-am`, but not always.
That's why we have both commands.

-  As for the order: it matters. The main advantage of using `git-am` is debugging. For instance, think of a CVE requiring 10 patches, and the 9th one causing a conflict. The user can go to the worktree and check the conflicting lines and see the current patch context. Whereas with `patch` alone the user would be overwhelmed with the previous 8 patches code in the git workspace. And lastly, fixing the conflicts with `git-am` is just a matter of changing a few lines and `git-am --continue`.

It's true though that the code in klp-build becomes more bloated by using `git-am`, but I think it's worth it. What do you think?

vmezzela replied:
Oh, I never seen it that way: when a patch failed to apply I'd usually check if we missed another one rather than adjusting the tree manually.

Does anyone  adjust it manually?

vmezzela commented on klpbuild/plugins/extract.py:
```diff
@@ -380,32 +380,74 @@ def group_equal_files(lp_name, working_cs):
+
+
+def apply_all_patches(lp_name, cs):
+>>>>>>> Stashed changes
```
leftover

fgyanz replied:
~~humm I think you are using and outdated version.~~

vmezzela replied:
Same as previous comment

fgyanz replied:
My bad, rebase chaos. Fixed

marcosps commented:
@fgyanz one thing that ocurred to me when I was rebasing my symbol-validation branch on top of the automation: the `--archs` now have no more usage. Even if we force an arch, but still set a `--conf`, `klp-build` will check the affected archs by using that `CONF`.

We have some tests that are using `--arch x86_64`, let's say, and using a config like `CONFIG_FS_PROC`, but klp-build sets the archs to all of them, since `CONFIG_FS_PROC`. To fix that, if we really want to create a fix for only one arch, in this test, I changed the config to `CONFIG_X86`, just in case.

We don't need to do anything for it right now, it's just a thing that I found, and I believe that we can finally remove the `--archs`, since it's inferred by the config right now.

vmezzela commented:
Hi,

> We don't need to do anything for it right now, it's just a thing that I found, and I believe that we can finally remove the `--archs`, since it's inferred by the config right now.

I think is better if we keep it. I know that we can change the config, but that's a bit hacky. What do you think?

fgyanz commented:
I agree with @vmezzela here, for the same reason. '--arch' is more user-friendly than some obscure config value that might  change between kernel versions. 

`We have some tests that are using --arch x86_64, let's say, and using a config like CONFIG_FS_PROC, but klp-build sets the archs to all of them, since CONFIG_FS_PROC. To fix that, if we really want to create a fix for only one arch, in this test, I changed the config to CONFIG_X86, just in case.`

Unless there's a bug, the expected behavior should not be that. If we set the config to  `CONFIG_FS_PROC` and the archs to `x86_64`, the affected codestreams should be within the intersection of `CONFIG_FS_PROC & --archs`. Meaning, all the codestreams that support CONFIG_FS_PROC in x86_64 arch.

marcosps commented:
> I agree with @vmezzela here, for the same reason. '--arch' is more user-friendly than some obscure config value that might change between kernel versions.
> 
> `We have some tests that are using --arch x86_64, let's say, and using a config like CONFIG_FS_PROC, but klp-build sets the archs to all of them, since CONFIG_FS_PROC. To fix that, if we really want to create a fix for only one arch, in this test, I changed the config to CONFIG_X86, just in case.`
> 
> Unless there's a bug, the expected behavior should not be that. If we set the config to `CONFIG_FS_PROC` and the archs to `x86_64`, the affected codestreams should be within the intersection of `CONFIG_FS_PROC & --archs`. Meaning, all the codestreams that support CONFIG_FS_PROC in x86_64 arch.

Sure, we can look at this later than. Just saying that, in my tests, this wasn't working as it was in the past. But I agree, one thing at a time. This won't hurt having it set like this for now.

fgyanz committed:
klplib/utils: Return basename in fix_mod_string

The modules are internally addressed by their full path for
convinience and safety, but that should not be the case in
generated source code. Third-party tools and kernel components
only need the modules basename. When adapting the modules to
string format, strip out the dirname.

Signed-off-by: Fernando Gonzalez <fernando.gonzalez@suse.com>

vmezzela commented on klpbuild/plugins/extract.py:
```diff
@@ -380,32 +380,79 @@ def group_equal_files(lp_name, working_cs):
+
+
+def apply_patch(patch, sdir):
+    # Try to apply the patch first with git-am.
```
Can we have a `logging.info()` here? A bit of time is being spent here and I don't like `klp-build` being totally silent for so long :D

fgyanz replied:
sure, done!

vmezzela commented:
Note: if I have kgraft-patches currently checked out on one of the livepatch branches, then the `commit` functionality wouldn't fully work. I don't know how this needs to be handled. But it was worth to mention it:

```
╭─vmezzela@localhost ~/Work/kgraft-patches ‹bsc1249537_15.7rtu0-4_15.7u0-4›
╰─$ klp-build commit -n bsc1249537 --force
Commiting livepatch to kgraft...
Failed to delete branch: +bsc1249537_15.7rtu0-4_15.7u0-4: error: branch '+bsc1249537_15.7rtu0-4_15.7u0-4' not found

Livepatch 'bsc1249537_6.0rtu2-12_6.0u2-11_15.6rtu5-14_15.6u6-15' commited
Livepatch 'bsc1249537_15.7rtu0-4_15.7u0-4' commited
Livepatch 'bsc1249537_16.0rtu0_16.0u0' commited
```

fgyanz commented:
That's true, but the whole point of having the worktree is to not modify the main repository. Should klp-build fail to delete the branch, it's then up to the user to fix it.

> Failed to delete branch: +bsc1249537_15.7rtu0-4_15.7u0-4: error: branch '+bsc1249537_15.7rtu0-4_15.7u0-4' not found

What's up with this error message? Oo and the `+bsc1249537_15.7rtu0-4_15.7u0-4` !!

vmezzela commented:
> That's true, but the whole point of having the worktree is to not modify the main repository. 

Indeed

> Should klp-build fail to delete the branch, it's then up to the user to fix it.

Agree

> 
> > Failed to delete branch: +bsc1249537_15.7rtu0-4_15.7u0-4: error: branch '+bsc1249537_15.7rtu0-4_15.7u0-4' not found
> 
> What's up with this error message? Oo and the `+bsc1249537_15.7rtu0-4_15.7u0-4` !!

It's the reason why I posted the comment :D 
Should we gracefully exit if that happens? Or at least inform the user what's happening and how to handle it (e.g. checkout to a different branch)

vmezzela commented on klpbuild/plugins/push.py:
```diff
@@ -71,62 +94,29 @@ def create_lp_package(osc, lp_name, i, total, cs):
-
-    logging.info("(%s/%s) pushing %s using branches %s/%s...",
-                 i, total, cs.full_cs_name(), product_branch, str(branch))
+    push_branch = create_push_branch(cs, branch)
```
In the future, it might be good to have an option to use as a fallback when this strategy doesn't work because of conflicts. Currently, when it happens I temporarily adjust the code so that it can at least push what I need and then I revert the change.. but that's not optimal.

fgyanz replied:
Makes sense. I added the original livepatch branch as a fallback. It should be enough to catch compilation issues, but for nothing else. Not sure if you had something else in mind though. Is there a strategy other than rebasing that we can use as fallback?

vmezzela commented on klpbuild/klplib/kgraft.py:
```diff
@@ -0,0 +1,110 @@
+    if err.stderr or not err.stdout:
+        return None
+
+    return [re.sub(r"[\s\*]",'', l) for l in err.stdout.splitlines()]
```
```suggestion
    return [re.sub(r"[\s\*+]",'', l) for l in err.stdout.splitlines()]
```


Also this case need to be handled. Otherwise when in the main `kgraft-patches` repository you have one of this branch checked out, this function will return `+bsc12345_*` rather than `bsc12345`.

fgyanz replied:
done, thanks a lot!

marcosps reviewed (approved):
LGTM!

vmezzela commented on klpbuild/plugins/push.py:
```diff
@@ -46,13 +53,36 @@ def create_prj_meta(cs):
+                        push_branch, lp_branch)
+        push_branch = lp_branch
+
+    return push_branch
```
Sorry, what I meant was to have an actual option like `--no-rebase` that skips the rebase. Because it needs to be done only when the user decides so. This of course can't work with codestreams groups but only with one codestream that has been previously manually rebased properly. 

The drawback of having an automatic fallback is that if the rebase fails and the fallback path is taken, most likely it would fail the build, because the user did not rebase the branch in advance manually..

What do you think?

fgyanz replied:
> Sorry, what I meant was to have an actual option like --no-rebase that skips the rebase. Because it needs to be done only when the user decides so. This of course can't work with codestreams groups but only with one codestream that has been previously manually rebased properly.

Ah I see! Yes, it makes sense. But for the future, indeed :D 

> The drawback of having an automatic fallback is that if the rebase fails and the fallback path is taken, most likely it would fail the build, because the user did not rebase the branch in advance manually..

Correct, that's expected. But I see two advantages for having this:
- It will attempt to compile the livepatch files. And although it will never fully succeed, it's good enough for catching some errors (like typical C mistakes). Of course, it's not meant for building livepatches, but it might be good enough during development if for whatever reason the product branches are broken.

- Sometimes only one product branch is conflicting or broken, in such case klp-build will currently crash due to the failed rebased. With this changes, it will at least continue pushing the rest of branches. 

Then again, this change does not fully address the case in your comment. But anyway, it might be somewhat useful :)

vmezzela commented:
LGTM!
